number_theory/numeric.h

```diff
@@ -43,6 +45,54 @@ std::pair<T, T> exgcd(T a, T b) {
   return std::make_pair(xa, ya);
 }
 
+// Accumulates values according to the binary representation of |binary|.
+// For each bit of |binary|, from lower to higher, it calls |update| to
+// accumulate the values. The initial value is |init|.
+template <class T, class U>
+U binary_accumulate(T binary,
+                    const U &init,
+                    std::function<void(bool, U &)> update) {
+  static_assert(
+      std::numeric_limits<T>::is_integer && !std::numeric_limits<U>::is_signed,
+      "binary_accumulate argument |n| must be an unsigned integer");
+
+  T current = binary;
+  U result = init;
+  while (current != 0) {
+    bool bit = (current % 2) != 0;
+    current /= 2;
+    update(bit, result);
+  }
+  return result;
+}
+
+// Computes the value of |base| raised to an integer power |exponent|.
+// The time complexity is O(log |exponent|).
+template <class T, class U>
+T ipow(const T &base, U exponent) {
+  static_assert(std::numeric_limits<U>::is_integer,
```

Comment:
I think we should return a floating point type when the `exponent` is negative. However, this is called `ipow` for integers only. Maybe, we should just disallow negative exponents here?

Reply:
see comments above